### PR TITLE
Service logs

### PR DESCRIPTION
**Pull request by granders**
* Updated services and tests to reflect refactoring in ducktape service-logs branch.
* Refactored PerformanceService to inherit from BackgroundThreadService.
* Broke up core.py into multiple modules.
* Broke up KafkaBenchmark into several small test classes.
* Added very simple run_tests.py script with the intention of executing this as a jenkins job.

### CODE REVIEW

File: aws/aws-access-keys-commands [hunk 1]
```diff
@@ -1,3 +1,3 @@
-export AWS_ACCESS_KEY=`curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/ducttape-master | grep AccessKeyId | awk -F\" '{ print $4 }'`
-export AWS_SECRET_KEY=`curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/ducttape-master | grep SecretAccessKey | awk -F\" '{ print $4 }'`
-export AWS_SESSION_TOKEN=`curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/ducttape-master | grep Token | awk -F\" '{ print $4 }'`
+export AWS_ACCESS_KEY=`curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/$AWS_IAM | grep AccessKeyId | awk -F\" '{ print $4 }'`
```
Comment (ewencp) [inline aws/aws-access-keys-commands]: I tried googling AWS_IAM to see if this is a predefined environment variable but couldn't find anything. Is there a reference for this, or is this supposed to be provided by the user?

File: muckrake/services/hadoop.py [hunk 2]
```diff
@@ -14,231 +14,10 @@
 
 from ducktape.services.service import Service
 
-import time, re
-from .schema_registry_utils import SCHEMA_REGISTRY_DEFAULT_REQUEST_PROPERTIES
-from .kafka_rest_utils import KAFKA_REST_DEFAULT_REQUEST_PROPERTIES
+import time
```
Comment (ewencp) [inline muckrake/services/hadoop.py]: Any logs to be specified for the classes in this file? There's a bunch of references to /mnt/logs, maybe we can generate at least a reasonable list as a starting point by running one of the tests and `ls`ing that directory?

File: muckrake/services/hadoop.py [hunk 3]
```diff
@@ -286,6 +68,15 @@ def start(self):
                 self.start_datanode(node)
             time.sleep(5)  # wait for start up
 
+    def stop_node(self, node):
+        pids = list(node.account.ssh_capture("ps ax | grep java | grep -v grep | awk '{print $1}'"))
```
Comment (ewencp) [inline muckrake/services/hadoop.py]: Something to maybe file a bug for in ducktape -- we should probably have reusable versions of stuff like this? Actually, a DaemonService that tracks real pids would probably be ideal, but for cases like this, not having them grab the PID list and kill the processes themselves would be better.

File: muckrake/services/zookeeper_service.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2015 Confluent Inc.
```
Comment (ewencp) [inline muckrake/services/zookeeper_service.py]: Nit: If we can avoid underscores by stripping unnecessary suffixes (like "_service" in this filename), I think that's better :)

File: muckrake/tests/camus_test.py [hunk 5]
```diff
@@ -69,16 +67,15 @@ def __init__(self, test_context):
 
         self.num_camus_perf = 1
 
-    def min_cluster_size(self):
-        return self.num_camus_perf + super(CamusHadoopV2Test, self).min_cluster_size()
+        self.services['camus_perf'] = CamusPerformanceService(
+            self.service_context(self.num_camus_perf),
+            self.kafka, self.hadoop, self.schema_registry, self.rest, settings={})
+        self.camus_perf = self.services['camus_perf']
```
Comment (ewencp) [inline muckrake/tests/camus_test.py]: Two things. First, a tip -- you can chain assignments in Python:

```
self.camus_perf = self.services['camus_perf'] = CamusPerformanceService(...)
```

Second, this makes me wonder if explicitly maintaining a separate services list is annoying for the user. I was thinking one thing we could do is just scan the fields of the object for instances of Service, although it'd have to be a deep traversal since a simple 1-level search wouldn't find something created like this:

```
self.producers = [ProducerPerformanceService(...) for x in range(3)]
```

File: muckrake/tests/kafka_benchmark_test.py [hunk 6]
```diff
@@ -27,164 +28,257 @@ def __init__(self, test_context):
...
         if False:
-            msgs_default = 1000000
-            msgs_large = 10000000
-            msg_size_default = 100
-            batch_size = 8*1024
-            buffer_memory = 64*1024*1024
-            msg_sizes = [10, 100, 1000, 10000, 100000]
-            target_data_size = 128*1024*1024
-            target_data_size_gb = target_data_size/float(1024*1024*1024)
+            # These settings will work in the default local Vagrant VMs, useful for testing
```
Comment (ewencp) [inline muckrake/tests/kafka_benchmark_test.py]: These look like they got inverted? The 50000000 version is not being run by default anymore? If we want to swap them for now so the tests run in the local VMs, that's fine, but at a minimum, the comment is mixed up.

File: muckrake/tests/kafka_benchmark_test.py [hunk 7]
```diff
@@ -27,164 +28,257 @@ def __init__(self, test_context):
...
-        # PRODUCER TESTS
+
+def throughput(perf):
+    """Helper method for computing throughput after running a performance service."""
+    aggregate_rate = sum([r['records_per_sec'] for r in perf.results])
+    aggregate_mbps = sum([r['mbps'] for r in perf.results])
+    return "%f rec/sec (%f MB/s)" % (aggregate_rate, aggregate_mbps)
+
+# PRODUCER TESTS
+class SingleProducerNoReplication(KafkaBenchmark):
```
Comment (ewencp) [inline muckrake/tests/kafka_benchmark_test.py]: Interesting that these are split up now. I suppose one benefit is that they are now truly isolated, and with a parallel test runner can be run in parallel. One drawback is that they now all have to separately pull up a Kafka cluster, which means serial test running will take a lot longer (and there are quite a few of these).

File: muckrake/tests/native_vs_rest_performance_test.py [hunk 8]
```diff
@@ -16,38 +16,44 @@
...
-        msg_size = 100
-        batch_size = 8196
-        acks = 1 # default for REST proxy, which isn't yet configurable
-        # These settings will work in the default local Vagrant VMs, useful for testing
         if False:
+            msgs = 50000000
+            msg_size = 100
+            batch_size = 8196
+            acks = 1 # default for REST proxy, which isn't yet configurable
+            # These settings will work in the default local Vagrant VMs, useful for testing
```
Comment (ewencp) [inline muckrake/tests/native_vs_rest_performance_test.py]: Same inversion of AWS/local settings here, but looks like the comment got mixed up at the bottom instead of being put at the top of the else block.

File: muckrake/tests/native_vs_rest_performance_test.py [hunk 9]
```diff
@@ -55,38 +61,45 @@ def __init__(self, test_context):
...
-        acks = 1 # default for REST proxy, which isn't yet configurable
-        nthreads = 1 # not configurable for REST proxy
-        # These settings will work in the default local Vagrant VMs, useful for testing
         if False:
+            msgs = 5000000
+            msg_size = 100
+            batch_size = 8196
+            acks = 1 # default for REST proxy, which isn't yet configurable
+            nthreads = 1 # not configurable for REST proxy
+            # These settings will work in the default local Vagrant VMs, useful for testing
```
Comment (ewencp) [inline muckrake/tests/native_vs_rest_performance_test.py]: Another inversion of AWS/local comment.

File: run_tests.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    print "Running %s" % cmd
+    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    for line in iter(proc.stdout.readline, ''):
+        print line
+    
+    output, err = proc.communicate()
+    print output
+    return proc.returncode
+
+def main():
```
Comment (ewencp) [inline run_tests.py]: All the tests? From scratch?! In one command!?! Ooh la la!

Seriously though, really happy to see it converging on this! :+1:

File: muckrake/tests/mini_test.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ducktape.tests.test import Test
+
+from muckrake.services.zookeeper_service import ZookeeperService
+from muckrake.services.kafka import KafkaService
+
+
+class MiniTest(Test):
```
Comment (ewencp) [inline muckrake/tests/mini_test.py]: What is this test for? Seems like a subset of everything_runs_test.py and doesn't have any subclasses?

**Comment by ewencp** [comment]
@granders So close! Ran the full test suite, but ran into a couple of problems. I think they should be easy to fix:

```
===========================
FAIL: <class 'muckrake.tests.kafka_benchmark_test.ProducerMultipleMessageSize'>
    There are not enough nodes available in the cluster to run this test. Needed: 9, Available: 7
Traceback (most recent call last):
  File "/Users/ewencp/confluent/ducktape.git/ducktape/tests/runner.py", line 52, in run_all_tests
    self.run_single_test(test)
  File "/Users/ewencp/confluent/ducktape.git/ducktape/tests/runner.py", line 69, in run_single_test
    (test.min_cluster_size(), self.cluster.num_available_nodes()))
RuntimeError: There are not enough nodes available in the cluster to run this test. Needed: 9, Available: 7
```

Can you just verify this one is because the number of nodes required due to specifying all the services together even if they are not running in parallel? Or did something else change? I could swear I only needed 7 nodes before.


```
===========================
FAIL: <class 'muckrake.tests.native_vs_rest_performance_test.NativeVsRestConsumerPerformance'>
    local variable 'acks' referenced before assignment
Traceback (most recent call last):
  File "/Users/ewencp/confluent/ducktape.git/ducktape/tests/runner.py", line 52, in run_all_tests
    self.run_single_test(test)
  File "/Users/ewencp/confluent/ducktape.git/ducktape/tests/runner.py", line 62, in run_single_test
    test = get_test_case(test_class, self.session_context)
  File "/Users/ewencp/confluent/ducktape.git/ducktape/tests/runner.py", line 44, in get_test_case
    return test_class(test_context)
  File "/Users/ewencp/confluent/muckrake.git/muckrake/tests/native_vs_rest_performance_test.py", line 79, in __init__
    settings={'batch.size':batch_size, 'acks': acks}
UnboundLocalError: local variable 'acks' referenced before assignment
```

Looks like one of the acks configuration lines got dropped from one of the conditional blocks?

File: run_tests.py (hunk 10 above)
Comment (granders) [inline run_tests.py]: It's not pretty, but it kinda works ;)

File: muckrake/tests/mini_test.py (hunk 11 above)
Comment (granders) [inline muckrake/tests/mini_test.py]: This isn't actually necessary for anything - it just provided me with a way to help test out docker, the run_tests.py script etc - faster to iterate when only two nodes were involved.

**Comment by granders** [comment]
9 nodes is currently necessary - one unfortunate consequence of the current allocation approach using the ServiceRegistry class. This test runs 5 or 6 of the same service sequentially and registers them all with the ServiceRegistry in __init__ before running the test, which means the test runner allocates a lot of extra nodes.

File: aws/aws-access-keys-commands (hunk 1 above)
Comment (granders) [inline aws/aws-access-keys-commands]: I wasn't sure the best approach here. The jenkins setup required use of a different IAM identity, so I invented this variable but haven't documented it.

File: muckrake/services/hadoop.py (hunk 2 above)
Comment (granders) [inline muckrake/services/hadoop.py]: Ah good point, this is an oversight.

**Comment by ewencp** [comment]
Ah, splitting up the performance test now makes a lot more sense -- that's another one that had a ton of services but they rarely ran in parallel.

File: aws/aws-access-keys-commands (hunk 1 above)
Comment (ewencp) [inline aws/aws-access-keys-commands]: Maybe add a default so it at least doesn't fail in weird ways if you don't have it defined? Presumably without the default the README instructions would currently be broken?

File: muckrake/tests/kafka_benchmark_test.py (hunk 6 above)
Comment (granders) [inline muckrake/tests/kafka_benchmark_test.py]: Yeah I intentionally inverted it but neglected the comment.

File: aws/aws-access-keys-commands (hunk 1 above)
Comment (granders) [inline aws/aws-access-keys-commands]: Added default.

File: muckrake/services/hadoop.py (hunk 3 above)
Comment (granders) [inline muckrake/services/hadoop.py]: Actually remoteaccount has a kill_process method.
Updated hadoop.py to use this

File: muckrake/services/zookeeper_service.py (hunk 4 above)
Comment (granders) [inline muckrake/services/zookeeper_service.py]: changed

File: muckrake/tests/kafka_benchmark_test.py (hunk 6 above)
Comment (granders) [inline muckrake/tests/kafka_benchmark_test.py]: Tweaked to clarify

File: muckrake/tests/kafka_benchmark_test.py (hunk 7 above)
Comment (granders) [inline muckrake/tests/kafka_benchmark_test.py]: Yeah I didn't think this was much of a problem for overnights.

File: muckrake/tests/native_vs_rest_performance_test.py (hunk 8 above)
Comment (granders) [inline muckrake/tests/native_vs_rest_performance_test.py]: tweaked

**Comment by granders** [comment]
@ewencp 
I pushed changes that hit your comments.

The service registration now happens automatically when service instances are created, so the test code is a little less noisy.

The hadoop and camus tests are now broken however, and I haven't figured out why - seems to hang on 'formatting namenode'

File: muckrake/services/hadoop.py (hunk 2 above)
Comment (granders) [inline muckrake/services/hadoop.py]: I added the base /mnt/logs directory with default no-collect behavior.

File: aws/aws-example-Vagrantfile.local [hunk 12]
```diff
@@ -5,4 +5,4 @@ ec2_keypair_name = 'confluent'
 ec2_keypair_file = '../confluent.pem'
 ec2_security_groups = ['ducttape-insecure']
 ec2_region = 'us-west-2'
-ec2_ami = "ami-af200d9f"
+ec2_ami = "ami-fdc2f6cd"
```
Comment (ewencp) [inline aws/aws-example-Vagrantfile.local]: Is this change intentional? I see this is our AMI and it sounds like it's just had vagrant provision run on it already, is the plan to just update this periodically?

File: muckrake/services/kafka.py [hunk 13]
```diff
@@ -17,28 +17,27 @@
...
+        :type context
         :type zk: ZookeeperService
         :type topics: dict
         """
-        super(KafkaService, self).__init__(service_context)
+        super(KafkaService, self).__init__(context, num_nodes)
         self.zk = zk
         self.topics = topics
+        self.logs = {
+            "kafka_log": {
```
Comment (ewencp) [inline muckrake/services/kafka.py]: Definitely not critical, but these feel kinda verbose to specify. Another way to handle this is specify that one mode is the default, and then use a special type of object to specify when it should use the opposite setting. For example, if we said the default was to collect the log, we might do something like:

```
from ducktape.service import DisableDefaultCollection
self.logs = { 'kafka_log': '/mnt/kafka.log', 'kafka_data': DisableDefaultCollection('/mnt/kafka-logs') }
```

The naming isn't great in this example, but you get the gist -- make the common case easy and concise. This just requires a simple conditional in the test runner to handle properly. Like I said, not critical (and my solution is probably not the best one), but might be worth filing a bug on ducktape to remember to try to improve this.

File: muckrake/services/hadoop.py [hunk 14]
```diff
@@ -286,6 +78,13 @@ def start(self):
                 self.start_datanode(node)
             time.sleep(5)  # wait for start up
 
+    def stop_node(self, node):
+        node.account.kill_process(self, "java", clean_shutdown=False)
```
Comment (ewencp) [inline muckrake/services/hadoop.py]: It looks like this has an extra `self` parameter? Signature of `kill_process` is currently:

```
def kill_process(self, process_grep_str, clean_shutdown=True, allow_fail=False):
```

so the first param should be "java", the string to grep for.

**Comment by ewencp** [comment]
@granders The current Hadoop test failures seem to have to do with failure to cleanup properly. On a clean node they succeed the first time, but some namenodes are left running. Then the subsequent run fails to clean those leftover processes up and hangs trying to format the namenode.

I think I tracked it down to HDPService overriding `stop_node` and `clean_node` with `pass`, which means it doesn't execute any of the stop/clean code in `HDFSService`, and it looks like format_namenode just hangs if one of the processes/files that are supposed to be cleaned up by those methods are still running/on the filesystem.

Fixing those (and the `kill_process` issue mentioned elsewhere) fixes hadoop_test for me. If it's really a correct fix, it should work for you as well, even on nodes that have leftover processes still running (since cleanup was not working properly before).

File: aws/aws-example-Vagrantfile.local (hunk 12 above)
Comment (granders) [inline aws/aws-example-Vagrantfile.local]: Yep, this is a pre-provisioned ami. Until there's a better solution I guess periodic manual updates will have to suffice.

File: muckrake/services/hadoop.py (hunk 14 above)
Comment (granders) [inline muckrake/services/hadoop.py]: oops good catch

**Comment by granders** [comment]
@ewencp 
Yeah that was my bad - I forgot or didn't realize I missed those while refactoring. 

I added some checks in muckrake to make this kind of problem more immediately visible. This revealed a few more services that weren't cleaning up after themselves fully.

File: aws/aws-example-Vagrantfile.local (hunk 12 above)
Comment (ewencp) [inline aws/aws-example-Vagrantfile.local]: Ok, that's actually a totally reasonable solution, and probably pretty easy to maintain since we'd only want to change it when we add new dependencies.

Can you just add a couple of lines to the README mentioning that you can do this and how (even if it's just a quick description of the steps, not necessarily exact commands).

File: muckrake/services/kafka.py (hunk 13 above)
Comment (granders) [inline muckrake/services/kafka.py]: I don't have a better idea at the moment. Leaving as is but opened https://github.com/confluentinc/muckrake/issues/27

**Comment by granders** [comment]
@ewencp Ready again :)

File: aws/aws-example-Vagrantfile.local (hunk 12 above)
Comment (granders) [inline aws/aws-example-Vagrantfile.local]: updated README

**Comment by ewencp** [comment]
@granders LGTM. I didn't run all the tests again to completely verify, but did run a variety of them. I actually ran into an error in one:

```
FAIL: <class 'muckrake.tests.sr_kafka_broker_failover_test.KafkaBrokerCleanBounce'>

-------------------------------------------------------------------
Summary
Histogram of number of attempts needed to successfully register:
{1: 1858, 5: 3, 9: 1, 49: 1, 83: 1, -1: 8, 63: 1}
Attempted to register 1873 schemas.
Max registration attempts allowed: 100
Retry backoff: 0.020000 seconds
Successful: 1865/1873 = 0.995729
Validating that schemas reported as successful can be fetched by id...
Success.
Validating that number of reported successful registrations matches number of versions in subject...
Success.
Validating that registered schemas match fetched schemas...
Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Found discrepencies between registered schemas and fetched schemas (id, registered, fetched). [(-1, '{"fields": [{"type": "string", "name": "f34"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f33"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f295"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f294"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f296"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f294"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f670"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f669"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f671"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f669"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f1032"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f1031"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f1033"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f1031"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f1476"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f1475"}], "type": "record", "name": "myrecord"}')]Failure.
Validating that successfully registered ids and schemas match ids and schemas fetched by subject/version...
Success.
Validating successfully registered ids agains ids fetched by subject/schema...
Success.
-------------------------------------------------------------------

Traceback (most recent call last):
  File "/Users/ewencp/confluent/ducktape.git/ducktape/tests/runner.py", line 52, in run_all_tests
    self.run_single_test(test)
  File "/Users/ewencp/confluent/ducktape.git/ducktape/tests/runner.py", line 78, in run_single_test
    test.run()
  File "/Users/ewencp/confluent/muckrake.git/muckrake/tests/test.py", line 348, in run
    self.report_summary()
  File "/Users/ewencp/confluent/muckrake.git/muckrake/tests/test.py", line 173, in report_summary
    assert success, summary
AssertionError:
-------------------------------------------------------------------
Summary
Histogram of number of attempts needed to successfully register:
{1: 1858, 5: 3, 9: 1, 49: 1, 83: 1, -1: 8, 63: 1}
Attempted to register 1873 schemas.
Max registration attempts allowed: 100
Retry backoff: 0.020000 seconds
Successful: 1865/1873 = 0.995729
Validating that schemas reported as successful can be fetched by id...
Success.
Validating that number of reported successful registrations matches number of versions in subject...
Success.
Validating that registered schemas match fetched schemas...
Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Failed to fetch id -1. Found discrepencies between registered schemas and fetched schemas (id, registered, fetched). [(-1, '{"fields": [{"type": "string", "name": "f34"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f33"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f295"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f294"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f296"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f294"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f670"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f669"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f671"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f669"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f1032"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f1031"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f1033"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f1031"}], "type": "record", "name": "myrecord"}'), (-1, '{"fields": [{"type": "string", "name": "f1476"}], "type": "record", "name": "myrecord"}', '{"fields": [{"type": "string", "name": "f1475"}], "type": "record", "name": "myrecord"}')]Failure.
Validating that successfully registered ids and schemas match ids and schemas fetched by subject/version...
Success.
Validating successfully registered ids agains ids fetched by subject/schema...
Success.
-------------------------------------------------------------------
```

But it didn't happen again and it seems unlikely it was caused by any of these changes. Let me know if I should file that one as a separate bug. I also have all the relevant logs thanks to these changes :)

**Comment by granders** [comment]
Thanks for the reviews!
